### ISSUE TITLE
Questions about HRL methods

### ISSUE THREAD

**Issue by TJU-work** [Guardianzc/DISCOpen-MedBox-DialoDiagnosis#4]
Hello, these are the results of my train and test HRL methods on the mz4 data set, but it seems that there is something wrong with the results shown, which are very different from the data reported in the paper. May I ask how to Train and Test your HRL methods? I'm following the steps in the Readme. I'm looking forward to hearing from you!
By the way, the data printed out in each simulation of the test results are the same. Is this normal?

<img width="1161" alt="image" src="https://github.com/Guardianzc/DISCOpen-MedBox-DialoDiagnosis/assets/150174132/5d0bfb3a-a9f4-4945-9a49-a2b79058fedb">
<img width="1168" alt="image" src="https://github.com/Guardianzc/DISCOpen-MedBox-DialoDiagnosis/assets/150174132/a54ad5c2-28a4-48a0-bbe0-cb70b6b7027d">

<img width="1351" alt="image" src="https://github.com/Guardianzc/DISCOpen-MedBox-DialoDiagnosis/assets/150174132/d64a48c7-5445-46c0-be81-5a7779d31c45">

**Comment by Guardianzc** [comment]
Well looking at the results there seems to be a problem with the training process that makes the model predict disease directly without symptom prediction, and on mz4 even if the disease classifier is called directly for disease diagnosis does not result in such a low rate of correctness, maybe check if the disease classifier is working properly. Also, for the best current model on symptom and disease inference see https://github.com/lemuria-wchen/DxFormer

**Comment by TJU-work** [comment]
Yes, I have seen the performance of Dxformer, but I'm still hoping to reproduce the effect of HRL, which currently has the best performance in the reinforcement approach. Could you please tell me how to train and test the HRL model correctly? I have followed the train and test in the Readme before, but it seems to be wrong, as you said. Looking forward to hearing from you!

**Comment by Guardianzc** [comment]
Thank you for your interest in this work and this area of research. In DISCOpen-MedBox-DialoDiagnosis/OpenMedicalChatBox/HRL
/run.py, you can adjust the settings for the experiment。
From your training result, it looks like the disease classifier are not being trained correctly. Perhaps try adjusting the "warm-up" settings to pre-train the data first.
Because of the server migration recently, it is difficult to get the previous experimental data for the time being, and we will check the specific experimental settings after the migration is completed.
In addition, DxFormer is also a reinforcement learning based work, if you want to reproduce the latest RL-results, you can still use it as one of the references.

**Comment by Guardianzc** [comment]
Okay I found a possible cause of this problem.
In the _current_state_representation_ function in _OpenMedicalChatBox\HRL\dialogue_manager\dialogue_manager_hrl.py_ on line 391, we used different encoding for different datasets, resulting in different encodings when using mz4 dataset was incorrectly encoded for the input disease discriminator, resulting in incorrect training of the disease discriminator.
One possible way to do this is to add a new setting to the _"__init__"_ of _OpenMedicalChatBox\HRL\run.py_ when training real-world datasets such as mz4:.
`parser.add_argument("--data_type", dest="data_type", type=str,default='real',help="the type of dataset.")`
You can also modify this function for other datasets so that the current state is encoded correctly.
I hope this solves your problem.

**Comment by Guardianzc** [comment]
Thank you for your interest in this work and the field, I will upload the changed code once the server is recovered! :D
